modules/recursion/arrow-recursion-data/src/main/kotlin/arrow/recursion/typeclasses/Recursive.kt

```diff
@@ -51,33 +51,42 @@ interface Recursive<T, F> {
 
   fun <A, W> T.gcata(CW: Comonad<W>, dist: DistFunc<F, W, A>, alg: WAlgebra<F, W, A>): A {
     fun c(x: T): Eval<Kind<W, Kind<F, Kind<W, A>>>> = FF().run {
-      Eval.later {
-        dist(
-          x.projectT().map {
-            CW.run { c(x).value().map(alg).duplicate() }
-          }
-        )
-      }
+      dist(
+        x.projectT().map {
+          CW.run { c(it).map { it.map(alg).duplicate() } }
+        }
+      )
     }
     return CW.run { alg(c(this@gcata).value().extract()) }
   }
 
   fun <A, W> T.gfold(CW: Comonad<W>, dist: DistFunc<F, W, A>, alg: WAlgebra<F, W, A>): A =
     gcata(CW, dist, alg)
 
+  /*
   fun <A, B> T.zygo(alg: Algebra<F, B>, f: (Kind<F, Tuple2<B, A>>) -> A): A =
     gfold(Tuple2.comonad<B>(), {
       distZygo(FF(), alg, FF().run { it.map { it.fix().map { it.fix() } } })
     }, {
       f(FF().run { it.map { it.fix() } })
     })
+*/
 
-  fun <A> T.histo(alg: (Kind<F, Cofree<F, A>>) -> A): A =
-    gcata<A, CofreePartialOf<F>>(Cofree.comonad(), {
-      distHisto(FF(), FF().run { it.map { it.fix() } })
-    }, {
-      alg(FF().run { it.map { it.fix() } })
-    })
+  fun <A> T.histo(alg: (Kind<F, Cofree<F, A>>) -> A): A {
+    fun go(t: T): Cofree<F, A> = FF().run {
```

Comment:
The Android compiler may choke on this. Favor opening and closing the runs with the smallest sope possible.

Reply:
Wait what? I never used android, what is special about it? o.O

Reply:
It doesn't like when there are two nested `this`-dependent calls.

Reply:
They run their own compiler/optimizer on top of the regular kotlin one.

Reply:
Hmm not quite sure I understand. Where exactly does this nest this calls? `go` only ever uses FF's this context right?

Reply:
I mean just 

```
FF().run { t.projectT().map { go(it) } }.let { ... } 
```